// ks3/config.js
var path = require('path');
var version = require('./package.json').version
module.exports = {
	protocol:'http',
	baseUrl:'ks3-cn-beijing.ksyuncs.com',
	// 所有header设置的前缀
	// ks3: kss , amazon s3: amz
	prefix:'kss',
	// 返回的数据格式,默认为xml,因为aws返回的全是xml,
	// 不过鉴于nodejs对于json的喜爱,可以在config中设置为`json`
	// ['xml','json']
	dataType: 'xml',
	// 默认返回的是这个格式
	// 非必要,不传递也不会报错
	//contentType:'application/xml',
	contentType:'',
	// 设置通过`user-agent`头信息
	ua:'KS3_NodeJS' + '/' + version,
	// 简单文件上传的最大限制
	uploadMaxSize : 5*1024*1024*1024,
	// 分块上传的最小单位
	chunkSize:5*1024*1024,
	//文件分块上传阈值下线
	multipartUploadMinSize: 100 * 1024 * 1024, // 100 MB
	// 分块上传重试次数
	retries:20,
	// 上传文件信息缓存地址
	cachePath:path.join(__dirname,'./cache/files/'),
	// 区域
	region:'BEIJING',
	/**
	 * endpoint
	 * 对应关系详见：https://docs.ksyun.com/documents/6761
	 */
	ENDPOINT : {
		/**
		 * @deprecated
		 */
		HANGZHOU : 'kss.ksyuncs.com',
		AMERICA: 'ks3-us-west-1.ksyuncs.com',
		BEIJING : 'ks3-cn-beijing.ksyuncs.com',
		HONGKONG: 'ks3-cn-hk-1.ksyuncs.com',
		SHANGHAI: 'ks3-cn-shanghai.ksyuncs.com',
		GUANGZHOU: 'ks3-cn-guangzhou.ksyuncs.com',
		RUSSIA: 'ks3-rus.ksyuncs.com',
		SINGAPORE: 'ks3-sgp.ksyuncs.com',
		JR_BEIJING: 'ks3-jr-beijing.ksyuncs.com',
		JR_SHANGHAI: 'ks3-jr-shanghai.ksyuncs.com',
		GOV_BEIJING: 'ks3-gov-beijing.ksyuncs.com'
	},
	isDebug: false, // 调试
	setRegion: function(region) {
		const tmp = this.ENDPOINT[region];
		if (tmp) {
			this.baseUrl = tmp
			this.region = region
		} else {
			this.setEndpoint(region)
		}
	},
	resetRegion: function(){
		this.baseUrl = this.ENDPOINT.BEIJING;
		this.region = '';
	},
	setEndpoint: function (endpoint) {
		if (endpoint) this.baseUrl = endpoint
	}
}
